src/Agent.Listener/Configuration/Validators.cs

```diff
@@ -92,5 +95,24 @@ public static bool NTAccountValidator(string arg)
 
             return true;
         }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA2000:Dispose objects before losing scope", MessageId = "locationServer")]
+        public static async Task<bool> IsHostedServer(string serverUrl, VssCredentials credentials, ILocationServer locationServer)
+        {
+            // Determine the service deployment type based on connection data. (Hosted/OnPremises)
+            VssConnection connection = VssUtil.CreateConnection(new Uri(serverUrl), credentials);
+            await locationServer.ConnectAsync(connection);
```

Comment:
Why do we need to re-establish connection for this check? Could we use existing one - which agent uses?

Reply:
This is an already existing method. I just made it public and moved it to another file, and now we are not sure if this is the right place for it — maybe we can just [leave it where it was?](https://github.com/microsoft/azure-pipelines-agent/blob/b99fdc2a509ce8b63521c8ca74755c8b39583a9a/src/Agent.Listener/Configuration/ConfigurationManager.cs#L776) I have not found where to get the connection that the agent uses, and without calling the `ConnectAsync` method, there will be a thrown exception.

Reply:
Can we add some getter for the connection then?

Reply:
Moved logic for getting the connection data to the separate method